lib/spack/spack/util/environment.py

```diff
@@ -25,6 +25,27 @@
 import os
 
 
+def filter_system_paths(paths):
+    """
+    Filter system paths and return a filtered array.
+
+    Arguments:
+    paths -- array of paths
+    """
+    system_paths = ['/usr']
```

Comment:
You probably want to be a bit more specific:
  1. We care about `lib`, `lib64`, and `include` in system locations, and we want them stripped from `-I`, `-L`, and `-Wl,-rpath`.
  2. We also care about `bin` for `PATH`, but I think we still need system paths to be added to `PATH` at lower precedence than any `PATH` from a Spack package.

For (1), consider what happens when `Homebrew` is installed in /usr/local/Homebrew and you try to depend on a homebrew keg-only package.  It'll be stripped by these rules, but it's actually not in any system path.

So why not:

```python
system_paths = ['/', '/usr/', '/usr/local']
suffixes = ['lib', 'lib64', 'include']
system_dirs = sum(([os.path.join(p, s)
                    for s in suffixes] for p in system_paths), [])

def filter_system_paths(paths):
    return [p for p in paths if not any(p in s for s in system_dirs)]
```

which gives:
```
>>> filter_system_paths([
    '/usr/local/Cellar/gcc/5.3.0/lib',
    '/usr/local/lib',
    '/usr/local/include',
    '/usr/local/lib64',
    '/usr/local/opt/some-package/lib',
    '/usr/opt/lib',
    '/lib',
    '/include',
    '/opt/some-package/include',
])
['/usr/local/Cellar/gcc/5.3.0/lib', '/usr/local/opt/some-package/lib', '/usr/opt/lib', '/opt/some-package/include']
```

We can also later extract `system_paths` and stick it in a `config.yaml`.